server/src/main/java/org/opensearch/rest/DelegatingRestHandler.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.rest;
+
+import org.opensearch.client.node.NodeClient;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Delegating RestHandler that delegates all implementations to original handler
+ *
+ * @opensearch.api
+ */
+public class DelegatingRestHandler implements RestHandler {
```

Comment:
@cwperks as far as I can tell, this class is not used by OpenSearch core and should not be the part of the API. If it is need, please make it as part of the API (plugins or/and extensions)

Reply:
Would this make sense as an inner class of ActionPlugin? The security plugin wraps RestHandlers using this extension point: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/ActionPlugin.java#L151-L153

Currently when the security plugin is installed, there is a bug where its not properly delegating implementations to the wrapped handler.

As a result this [call in RestController](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/rest/RestController.java#L313) is not properly being delegated to and instead falling back on the default implementation in RestHandler.

Reply:
You basically need a mechanism to wrap around the `RestHandler`s before they come into the `ActionRegistry`, it that correct?

Reply:
The [SecurityRestFilter](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/filter/SecurityRestFilter.java#L133) in the security plugin wraps every RestHandler using an anonymous class. The anonymous class is not delegating implementations to the wrapped handler and instead using the defaults from the RestHandler interface instead. 

[RestSearchAction](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/rest/action/search/RestSearchAction.java#L404-L407) is an example RestHandler that overrides the default implementation of `allowsUnsafeBuffers` from the `RestHandler` interface.

Since the security plugin is not delegating other implementations to the wrapped RestHandler, its currently returning `false` which is the [default implement in RestHandler](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/rest/RestHandler.java#L81-L83) instead of wrapped handler's overridden implementation which would return `true` instead.

Reply:
> Since the security plugin is not delegating other implementations to the wrapped RestHandler, its currently returning false which is the [default implementation in RestHandler](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/rest/RestHandler.java#L81-L83) instead of the wrapped handler's overridden implementation which would return true instead

I have a silly question, sorry about that, why  [SecurityRestFilter](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/filter/SecurityRestFilter.java#L133) could not delegate to the original handler when it creates an anonymous class?

Reply:
That's what the companion security PR will do https://github.com/opensearch-project/security/pull/3517. (See PR description)

Instead of the anonymous class it will create a class that extends `DelegatingRestHandler` from this PR. That way it will ensure that when the SecurityRestFilter wraps the original RestHandler, all of the other implementations are delegated to the original handler as well.

Reply:
So just move this class to `security` plugin than, it is not needed in core

Reply:
It could be done wholly within the security plugin, but I thought it would be better to introduce the DelegatingRestHandler here so that if someone updated the RestHandler interface they would know to also update this class as well.

If someone changed the interface, it might not be immediately apparent in the Security plugin unless a default implementation wasn't provided and the security plugin started to fail to compile.

Reply:
`RestHandler` is public API and won't be changed in a backward incompatible way, `DelegatingRestHandler` will not solve the problem if added methods are `default` anyway. Please move this change to `security` plugin, thanks a lot for staying with me, regarding catching API changes, I think we could have a generic mechanisms to do that during build time.

Reply:
Sounds good. I'll close this PR and roll this into the security PR.